lib/smooth/application/display/DisplaySpi.cpp

```diff
@@ -59,87 +58,146 @@ namespace smooth::application::display
+        res &= send_data(&value, 1);
 
         return res;
     }
 
-    // Reset the display
-    void ILI9341::reset_display()
+    // Hardware Reset
+    void DisplaySpi::hw_reset(bool active_low, int active_time, int delay_time)
     {
-        reset_pin.set(PIN_LOW);
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
-        reset_pin.set(PIN_HIGH);
-        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+        if (active_low)
+        {
+            reset_pin.set(PIN_LOW);     // force the display chip to reset
+            std::this_thread::sleep_for(std::chrono::milliseconds(active_time));
+            reset_pin.set(PIN_HIGH);
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+        }
+        else
+        {
+            reset_pin.set(PIN_HIGH);    // force the display chip to reset
+            std::this_thread::sleep_for(std::chrono::milliseconds(active_time));
+            reset_pin.set(PIN_LOW);
+            std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+        }
+    }
+
+    // Software Reset
+    bool DisplaySpi::sw_reset(int delay_time)
+    {
+        bool res = send_cmd(static_cast<uint8_t>(LcdCmd::SWRESET));
+        std::this_thread::sleep_for(std::chrono::milliseconds(delay_time));
+
+        return res;
     }
 
     // Send initialize sequence of commands and data to display
-    bool ILI9341::send_init_cmds()
+    bool DisplaySpi::send_init_cmds(const display_init_cmd_t* init_cmds, size_t length)
     {
         bool res = true;
 
-        uint16_t seq_num = 0;
-
-        while (ili_init_cmds.at(seq_num).databytes != 0xff)
+        for (size_t index = 0; index < length; index++)
         {
-            // check if delay is needed; send only command then delay
-            if (ili_init_cmds.at(seq_num).databytes & 0x80)
+            // send command
+            res &= send_cmd(init_cmds[index].cmd);
+
+            // send data
+            uint8_t data_length = init_cmds[index].length;
+
+            if (data_length >= 1 && data_length < 0x80)
             {
-                res &= lcd_wr_cmd(ili_init_cmds.at(seq_num).cmd);
-                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                res &= send_data(init_cmds[index].data, data_length);
             }
-            else
+
+            // check to see if we have to delay before advancing to next sequence in ints_cmds
+            if (data_length >= 0x80)
             {
-                res &= lcd_wr_cmd(ili_init_cmds.at(seq_num).cmd);
-                res &= lcd_wr_data(ili_init_cmds.at(seq_num).data, ili_init_cmds.at(seq_num).databytes & 0x1F);
+                uint8_t delay_data_length = static_cast<uint8_t>(data_length & 0x7F);
+                size_t delay_time = init_cmds[index].data[delay_data_length];
+
+                // check delay_data_length to see if there is data to send before delaying
+                if (delay_data_length != 0)
```

Comment:
Using != isn't wrong, but `delay_data_length > 0` makes more sense in this context. Change it or leave it as-is.

Reply:
Done